lib/div.cpp

```diff
@@ -0,0 +1,457 @@
+
+  if (a_.dtype() != out_dtype) {
+    a_ = a_.to(out_dtype);
+  }
+  if (!a_.sizes().equals(a.sizes())) {
+    a_ = at::empty(a.sizes(), a_.options().dtype(out_dtype));
+  }
+  TORCH_CHECK(a_.sizes() == a.sizes(), "a_ size mismatch after broadcast.");
+
+  std::string kernel_name = use_integer_kernel ? "int_floordiv_kernel_" : "float_floordiv_kernel_";
+  const TritonJITFunction& f =
+      TritonJITFunction::getInstance((utils::get_triton_src_path() / "div.py").string(), kernel_name);
+
+  TORCH_CHECK(a.dim() >= 2, "Input tensor must be at least 2D for broadcasting");
+  int64_t M = a.size(0);
+  int64_t N = a.numel() / M;
+  int64_t tile_size = 1024;
+  int64_t total = M * N;
+  int num_blocks = (total + tile_size - 1) / tile_size;
+
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  CUstream raw_stream = static_cast<CUstream>(stream.stream());
+  c10::DeviceGuard guard(a_.device());
+
+  f(raw_stream, num_blocks, 1, 4, 2, 0, a.to(out_dtype), b.to(out_dtype), a_, M, N, tile_size);
+
+  return a_;
+}
+// ==================== main ====================
+at::Tensor floor_div_(at::Tensor& a_, const at::Tensor& b_) {
+  // ========== Tensor // Scalar ==========
+  if (a_.dim() > 0 && b_.dim() == 0) {
+    at::Tensor b_expand = at::full_like(a_, b_.item<double>());
+    bool is_int = c10::isIntegralType(a_.scalar_type(), true) && c10::isIntegralType(b_.scalar_type(), true);
+    return launch_floor_div_kernel_(a_, b_expand, is_int);
+  }
+
+  // ========== Tensor // Tensor ==========
+  bool is_int = c10::isIntegralType(a_.scalar_type(), true) && c10::isIntegralType(b_.scalar_type(), true);
+  return launch_floor_div_kernel_(a_, b_, is_int);
+}
+
+at::Tensor div_mode(const at::Tensor& a_,
+                    const at::Tensor& b_,
+                    const c10::optional<std::string>& rounding_mode) {
+  auto res = torch::broadcast_tensors({a_, b_});
+  at::Tensor a = res[0].contiguous();
+  at::Tensor b = res[1].contiguous();
+
+  at::ScalarType out_dtype = at::promote_types(a.scalar_type(), b.scalar_type());
+  at::Tensor out = at::empty(a.sizes(), at::TensorOptions().dtype(out_dtype).device(a.device()));
+
+  std::string kernel_name;
+  if (rounding_mode == "floor") {
+    kernel_name =
+        c10::isIntegralType(a.scalar_type(), true) ? "int_floordiv_kernel" : "float_floordiv_kernel";
+    ;
+  } else if (rounding_mode == "trunc") {
+    kernel_name = "trunc_divide_kernel";
+  } else if (!rounding_mode.has_value() || rounding_mode.value() == "none") {
+    kernel_name = "true_div_kernel";
+  } else {
+    TORCH_CHECK(false, "div_mode rounding_mode must be 'floor', 'trunc', or empty.");
+  }
+  const TritonJITFunction& f =
+      TritonJITFunction::getInstance((utils::get_triton_src_path() / "div.py").string(), kernel_name);
+
+  int64_t tile_size = 1024;
+  int64_t n = out.numel();
+  const unsigned int num_blocks = (n + tile_size - 1) / tile_size;
```

Comment:
On NVidia GPU grid.x's limit is 2^31 - 1. Though it's a super narrow chance to overflow, but using signed int is better here.